### PR TITLE
feat: Simplify error logging

### PR DESCRIPTION
**Pull request by JonathanWoollett-Light**
## Changes

Simplifies errors to always log their `Display` implementation and
always write their `Debug` implementation to `stderr`, this also
simplifies general logging procedure when a log target has not been
initialized, now all logs are written to `stdout`. Previously `error`
and `warn` level logs where written to `stderr` while `info`, `debug`
and `trace` level logs where written to `stdout`.

## Reason

Simpler is better (it also helps with https://github.com/firecracker-microvm/firecracker/pull/3971).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following
Developer Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

- [x] If a specific issue led to this PR, this PR closes the issue.
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] API changes follow the [Runbook for Firecracker API changes][2].
- [x] User-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.
- [x] New `TODO`s link to an issue.
- [x] Commits meet [contribution quality standards](https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md#contribution-quality-standards).

---

- [x] This functionality cannot be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

### CODE REVIEW

File: src/firecracker/tests/test_api_server.rs [hunk 1]
```diff
@@ -0,0 +1,42 @@
+// Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+const BINARY: &str = env!("CARGO_BIN_EXE_firecracker");
```
Comment (pb8o) [inline src/firecracker/tests/test_api_server.rs:4]: Please don't make integration tests in Rust. We already have the Python integration tests for that. 

Furthermore, this test seems to duplicate https://github.com/firecracker-microvm/firecracker/blob/main/tests/integration_tests/functional/test_api_server.py#L10

**Review by pb8o — CHANGES_REQUESTED** [review]

File: src/firecracker/tests/test_api_server.rs (hunk 1 above)
Comment (JonathanWoollett-Light) [reply src/firecracker/tests/test_api_server.rs:4]: I don't think the existing tests look at stdout and stderr. They only check for the presence of the error in the log target (and check the exit-code is non-zero but not a specific value).

Comment (pb8o) [reply src/firecracker/tests/test_api_server.rs:4]: > We have existing integration tests in https://github.com/firecracker-microvm/firecracker/blob/main/src/vmm/tests/integration_tests.rs. I found this the simplest test to specifically cover the change. I don't see a reason for not having this additional test?

To me that's still a unit test, as it does not use the firecracker binary. That's where I make the distinction.

File: src/firecracker/src/main.rs [hunk 2]
```diff
@@ -75,16 +75,25 @@ impl From<MainError> for ExitCode {
...
+struct MainExit(Result<(), MainError>);
+impl std::process::Termination for MainExit {
+    fn report(self) -> ExitCode {
+        if let Err(err) = self.0 {
+            error!("{err}");
+            eprintln!("Error: {err:?}");
+            ExitCode::from(err)
+        } else {
+            ExitCode::SUCCESS
+        }
```
Comment (ShadowCurse) [inline src/firecracker/src/main.rs:89]: I don't really like adding an additional struct just to print errors. I think you can just modify the main itself:
```rust
fn main() -> ExitCode {
    if let Err(err) = main_exec() {
        error!("{err}");
        eprintln!("Error: {err:?}");
        ExitCode::from(err)
    } else {
        ExitCode::SUCCESS
    }
}

```

Comment (JonathanWoollett-Light) [reply src/firecracker/src/main.rs]: Fixed.

**Review by ShadowCurse — APPROVED** [review]

**Review by pb8o — APPROVED** [review]

**Pull request merged**